src/write_gcm.c

```diff
@@ -2213,6 +2281,59 @@ static monitored_resource_t *monitored_resource_create_from_fields(
                                               num_labels);
 }
 
+static monitored_resource_t *parse_monitored_resource(char *metadata,
+    int metadata_size, const char *project_id) {
+  yajl_val node;
+  char errbuf[1024];
+  node = yajl_tree_parse(metadata, errbuf, sizeof(errbuf));
+  if (node == NULL) {
+    if (strlen(errbuf)) {
+      DEBUG("write_gcm: parse_monitored_resource %s", errbuf);
+    } else {
+      DEBUG("write_gcm: parse_monitored_resource: Not valid JSON response.");
+    }
+    goto error;
+  }
+  {
+    const char * type_path[] = {"type", (const char *) 0};
+    const char * labels_path[] = {"labels", (const char *) 0};
+    char *type;
+    yajl_val type_v = yajl_tree_get(node, type_path, yajl_t_string);
+    yajl_val labels_v = yajl_tree_get(node, labels_path, yajl_t_object);
+    if (type_v && YAJL_IS_STRING(type_v)) {
+      type = YAJL_GET_STRING(type_v);
+    } else {
+      ERROR("write_gcm: wg_parse_monitored_resource: %s not correctly defined.",
+            type_path[0]);
+      goto error;
+    }
+    size_t num_values;
+    if (labels_v && YAJL_IS_OBJECT(labels_v)) {
+      num_values = labels_v->u.object.len;
+    } else {
+      ERROR("write_gcm: wg_parse_monitored_resource: No such node %s.\n",
+            labels_path[0]);
+      goto error;
+    }
+    label_t labels[num_values];
+    for (int i = 0; i < num_values; i++) {
+      labels[i].key = *(labels_v->u.object.keys + i);
+      labels[i].value = YAJL_GET_STRING(*(labels_v->u.object.values + i));
+    }
+    monitored_resource_t *response =
+        monitored_resource_create_from_array(type, project_id, labels,
+                                             num_values);
+    yajl_tree_free(node);
+    sfree(metadata);
+    return response;
```

Comment:
If you really want to keep 'goto's (I don't like to use them), then how about:
* Replace 'error:' with 'finish:';
* Replace this line and yajl_tree_free(node) and sfree(metadata) with 'goto finish;'
* Above, where the type_path is not correctly defined, replace 'goto error;' with 'monitored_resource_t* response = NULL; goto finish;'?

This should eliminate duplication of the memory cleanup code and ensure it can't accidentally get out of sync.